drivers: spi: supporting Andes atcspi200 driver

kevinwang821020 opened the pull request:
Support andes atcspi200 driver, and use board adp_xc7k_ae350 for testing.
Also the driver's Kconfig has fit the current policy, the Kconfig is depend on status in device node (DT_HAS_XXXXX).

fkokosinski commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -0,0 +1,629 @@
+	sys_write32(int_msk, SPI_INTEN(dev));
+
+	if (!data->is_cmdaddr_mode) {
+		/* Start transfering */
```
`transfering` -> `transferring` (checkpatch complains about this typo)

kevinwang821020 replied:
Thanks for reviewing!!
I have fixed it.

fkokosinski commented on dts/bindings/spi/andestech.atcspi200.yaml:
```diff
@@ -0,0 +1,21 @@
+
+    num-cs:
+      type: int
+      required: false
```
IIRC `required` defaults to `false`, so there's no need to specify it if you don't want to make this field required.

kevinwang821020 replied:
Oh! I got it.
Thank you for your help, I have fixed it.

bbilas commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -0,0 +1,629 @@
+		SPI_DMA_CHANNEL(n, tx, TX, MEMORY, PERIPHERAL)		\
+	};								\
+	static void spi_atcspi200_cfg_##n(void);			\
+	static struct spi_atcspi200_cfg spi_atcspi200_dev_cfg_##n = {	\
```
```suggestion
	static const struct spi_atcspi200_cfg spi_atcspi200_dev_cfg_##n = {	\
```

kevinwang821020 replied:
Thanks for reviewing!

bbilas replied:
It seems that you have forgotten about that.

kevinwang821020 replied:
It my mistake, thanks for your remind.

bbilas commented on dts/bindings/spi/andestech.atcspi200.yaml:
```diff
@@ -0,0 +1,20 @@
+    reg:
+      required: true
+
+    num-cs:
```
Where is that being used?

kevinwang821020 replied:
Thanks for reviewing.

Originally, we just hoped to let users know the number of CS our spi driver support. And due to our spi IP provide dedicated CS(automatic control by hardware), so we only support one dedicated CS pin.

After seeing your suggestion, we also think that it is not very suitable to be a property just for the purpose of providing information, so we decided to remove it, thank you very much.

bbilas commented:
What about non-dedicated CS gpios? Is there any particular reason not to use them?

kevinwang821020 commented:
> What about non-dedicated CS gpios? Is there any particular reason not to use them?

kevinwang821020 commented:
> What about non-dedicated CS gpios? Is there any particular reason not to use them?

Thanks for your review.
Because the dedicated CS pin in our SPI IP is automatically controlled by hardware, it is more efficient.
So currently we tend to only support dedicated CS for our SPI driver.

bbilas commented:
> > What about non-dedicated CS gpios? Is there any particular reason not to use them?
> 
> Thanks for your review. Because the dedicated CS pin in our SPI IP is automatically controlled by hardware, it is more efficient. So currently we tend to only support dedicated CS for our SPI driver.

That might be a problem in case of troubles with routing dedicated CS GPIO on the PCB design. In that case, the user will not be able to use non-dedicated CS GPIO instead as it is done in the rest of the SPI drivers in the Zephyr.

bbilas commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -0,0 +1,629 @@
+/*
+ * Copyright (c) 2019 Andes Technology Corporation.
```
s/2019/2022

kevinwang821020 replied:
Thanks for reviewing.

I have fixed this incorrect date.

bbilas commented on drivers/spi/Kconfig.andes_atcspi200:
```diff
@@ -0,0 +1,12 @@
+# Kconfig Andes ATCSPI200 SPI configuration options
+#
+# Copyright (c) 2019 Andes Technology Corporation.
```
s/2019/2022

kevinwang821020 replied:
Thanks for reviewing.

I have fixed this incorrect date.

bbilas commented on dts/bindings/spi/andestech.atcspi200.yaml:
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (c) 2019, Andes Technology Corporation.
```
s/2019/2022

kevinwang821020 replied:
Thanks for reviewing.

I have fixed this incorrect date.

fkokosinski commented on boards/riscv/adp_xc7k_ae350/adp_xc7k_ae350_defconfig:
```diff
@@ -13,6 +13,9 @@ CONFIG_SERIAL=y
 
+# SPI driver options
+CONFIG_SPI=y
+
```
Don't y-select `CONFIG_SPI` by default. See this PR: https://github.com/zephyrproject-rtos/zephyr/pull/48574

kevinwang821020 replied:
OK! I got it!

Thank a lot for your remind

carlocaione commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -0,0 +1,629 @@
+#define REG_MCTRL       0x50
+#define REG_SLVST       0x60
+#define REG_SDCNT       0x64
+#define REG_CONFIG      0x7c
```
uhm, too many of them for the DT. These are just offsets from a base address defined into the DT so I think this is fine.

carlocaione commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -0,0 +1,629 @@
+#define REG_SDCNT       0x64
+#define REG_CONFIG      0x7c
+
+#define DEV_CFG(dev) ((const struct spi_atcspi200_cfg * const) (dev)->config)
+#define DEV_DATA(dev) ((struct spi_atcspi200_data *) (dev)->data)
```
We are removing `DEF_CFG` and `DEV_DATA` from drivers. Do not use it.

kevinwang821020 replied:
Thanks for reviewing.

I have remove the inappropriate usage.

carlocaione commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -0,0 +1,629 @@
+#define DEV_CFG(dev) ((const struct spi_atcspi200_cfg * const) (dev)->config)
+#define DEV_DATA(dev) ((struct spi_atcspi200_data *) (dev)->data)
+
+#define SPI_IDR(dev) (DEV_CFG(dev)->base + REG_IDR)
+#define SPI_TFMAT(dev) (DEV_CFG(dev)->base + REG_TFMAT)
+#define SPI_DIRIO(dev) (DEV_CFG(dev)->base + REG_DIRIO)
+#define SPI_TCTRL(dev) (DEV_CFG(dev)->base + REG_TCTRL)
+#define SPI_CMD(dev) (DEV_CFG(dev)->base + REG_CMD)
+#define SPI_ADDR(dev) (DEV_CFG(dev)->base + REG_ADDR)
+#define SPI_DATA(dev) (DEV_CFG(dev)->base + REG_DATA)
+#define SPI_CTRL(dev) (DEV_CFG(dev)->base + REG_CTRL)
+#define SPI_STAT(dev) (DEV_CFG(dev)->base + REG_STAT)
+#define SPI_INTEN(dev) (DEV_CFG(dev)->base + REG_INTEN)
+#define SPI_INTST(dev) (DEV_CFG(dev)->base + REG_INTST)
+#define SPI_TIMIN(dev) (DEV_CFG(dev)->base + REG_TIMIN)
+#define SPI_MCTRL(dev) (DEV_CFG(dev)->base + REG_MCTRL)
+#define SPI_SLVST(dev) (DEV_CFG(dev)->base + REG_SLVST)
+#define SPI_SDCNT(dev) (DEV_CFG(dev)->base + REG_SDCNT)
+#define SPI_CONFIG(dev) (DEV_CFG(dev)->base + REG_CONFIG)
```
Please remove the macros not actually used.

kevinwang821020 replied:
OK! I got it!

Thanks a lot!

carlocaione commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -0,0 +1,629 @@
+#define SPI_CONFIG(dev) (DEV_CFG(dev)->base + REG_CONFIG)
+
+/* Field mask of SPI transfer format register */
+#define REG_TFMAT_CPHA_OFFSET (0)
+#define REG_TFMAT_CPOL_OFFSET (1)
+#define REG_TFMAT_SLVMODE_OFFSET (2)
+#define REG_TFMAT_LSB_OFFSET (3)
+#define REG_TFMAT_MOSI_BIDIR_OFFSET (4)
+#define REG_TFMAT_DATA_MERGE_OFFSET (7)
+#define REG_TFMAT_DATA_LEN_OFFSET (8)
+#define REG_TFMAT_ADDR_LEN_OFFSET (16)
```
Pick one indentation style and stick to it (cfr with `REG_*`)

kevinwang821020 replied:
I have change the indentation style.
Thanks for your remind.

carlocaione commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -0,0 +1,629 @@
+#define REG_DIRIO_CS_OUTPUT_MSK (0x1 << REG_DIRIO_CS_OUTPUT_OFFSET)
+#define REG_DIRIO_CS_OUTPUT_EN_MSK (0x1 << REG_DIRIO_CS_OUTPUT_EN_OFFSET)
+
+/* Field mask of SPI transfer control register */
+#define REG_TCTRL_RD_TCNT_OFFSET (0)
+#define REG_TCTRL_DUMMY_CNT_OFFSET (9)
+#define REG_TCTRL_TOKEN_VAL_OFFSET (11)
+#define REG_TCTRL_WR_TCNT_OFFSET (12)
+#define REG_TCTRL_TOKEN_EN_OFFSET (21)
+#define REG_TCTRL_TRNS_MODE_OFFSET (24)
+#define REG_TCTRL_ADDR_FMT_OFFSET (28)
+#define REG_TCTRL_ADDR_EN_OFFSET (29)
+#define REG_TCTRL_CMD_EN_OFFSET (30)
```
Remove unneeded defines and move everything to an header file.

kevinwang821020 replied:
OK, I see!
Thanks for your remind.

carlocaione commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -0,0 +1,629 @@
+#define REG_STAT_TX_EMPTY_OFFSET (22)
+#define REG_STAT_TX_FULL_OFFSET (23)
+
+#define REG_STAT_SPI_ACT_MSK (0x1 << REG_STAT_SPI_ACT_OFFSET)
+#define REG_STAT_RX_NUM_MSK (0x1f << REG_STAT_RX_NUM_OFFSET)
+#define REG_STAT_RX_EMPTY_MSK (0x1 << REG_STAT_RX_EMPTY_OFFSET)
+#define REG_STAT_RX_FULL_MSK (0x1 << REG_STAT_RX_FULL_OFFSET)
+#define REG_STAT_TX_NUM_MSK (0x1f << REG_STAT_TX_NUM_OFFSET)
+#define REG_STAT_TX_EMPTY_MSK (0x1 << REG_STAT_TX_EMPTY_OFFSET)
+#define REG_STAT_TX_FULL_MSK (0x1 << REG_STAT_TX_FULL_OFFSET)
```
Use `GENMASK` when possible.

kevinwang821020 replied:
That's great!
Thanks for your suggestion.

carlocaione commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -0,0 +1,629 @@
+#define RX_NUM_STAT(dev) (sys_read32(SPI_STAT(dev)) & REG_STAT_RX_NUM_MSK)
+#define GET_TX_NUM(dev) (TX_NUM_STAT(dev) >> REG_STAT_TX_NUM_OFFSET)
+#define GET_RX_NUM(dev) (RX_NUM_STAT(dev) >> REG_STAT_RX_NUM_OFFSET)
+
```
empty space

kevinwang821020 replied:
OK, I see!
Thanks for your remind.

carlocaione commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -0,0 +1,629 @@
+	SET_MASK(SPI_TFMAT(dev), (data_len << REG_TFMAT_DATA_LEN_OFFSET));
+
+	/* Set SPI frame format */
+	if (config->operation & SPI_MODE_CPHA)
+		SET_MASK(SPI_TFMAT(dev), REG_TFMAT_CPHA_MSK);
+	else
+		CLR_MASK(SPI_TFMAT(dev), REG_TFMAT_CPHA_MSK);
```
Missing `{`

kevinwang821020 replied:
OK, I see!
Thanks for your remind.

carlocaione commented on drivers/spi/spi_andes_atcspi200.c:
```diff
@@ -0,0 +1,629 @@
+
+static int spi_transfer(const struct device *dev, uint32_t len)
+{
+
```
empty space

kevinwang821020 replied:
OK, I see!
Thanks for your remind.

kevinwang821020 committed:
drivers: spi: add Andes atcspi200 driver

Support the Andes atcspi200 spi driver

Signed-off-by: Kevin Wang <yunkai@andestech.com>

kevinwang821020 committed:
boards: riscv: adp_xc7k_ae350 support SPI driver

Enable spi node in dts
Disable flash node due to haven't implemented yet.
Add spi to the supporting device list in board yaml.

Signed-off-by: Kevin Wang <yunkai@andestech.com>

kevinwang821020 committed:
dts: bindings: spi: add andes spi driver

Add Andes SPI atcspi200 dts binding.
Remove not necessary property for spi node in soc dtsi.

Signed-off-by: Kevin Wang <yunkai@andestech.com>

kevinwang821020 committed:
tests: drivers: spi: spi_loopback: Add board's overlay

Add child node in spi node for testcase parsing compatible

Signed-off-by: Kevin Wang <yunkai@andestech.com>

kevinwang821020 commented:
In addition to modify according to your suggestions, due to the spi context API has been updated in v3.2, I also modified it according to the updated content.

Thanks a lot.

kevinwang821020 commented:
> > > What about non-dedicated CS gpios? Is there any particular reason not to use them?
> > 
> > 
> > Thanks for your review. Because the dedicated CS pin in our SPI IP is automatically controlled by hardware, it is more efficient. So currently we tend to only support dedicated CS for our SPI driver.
> 
> That might be a problem in case of troubles with routing dedicated CS GPIO on the PCB design. In that case, the user will not be able to use non-dedicated CS GPIO instead as it is done in the rest of the SPI drivers in the Zephyr.

Thanks for your suggestion.
It is a great point.
And we have let our driver support non-dedicated CS via gpio.

kevinwang821020 commented:
It seems like there are some problem when building and running qemu_x86.
But I thought there's nothing I can do in this PR to avoid them.
Is it reasonable to keep these errors here?

bbilas commented:
> It seems like there are some problem when building and running qemu_x86. But I thought there's nothing I can do in this PR to avoid them. Is it reasonable to keep these errors here?

It seems to be related to the https://github.com/zephyrproject-rtos/zephyr/pull/49646.

kevinwang821020 commented:
> > It seems like there are some problem when building and running qemu_x86. But I thought there's nothing I can do in this PR to avoid them. Is it reasonable to keep these errors here?
> 
> It seems to be related to the #49646.

Yes, it seems like the same issue.
Thanks a lot!
